docs/rfc/rfc-002-wasm-contract-parallelization.md

```diff
@@ -0,0 +1,37 @@
+# RFC 002: Wasm Contract Parallelization
+
+## Changelog
+
+- 2022-09-30: Initial Draft
+
+## Abstract
+
+This document is meant as an extension to RFC-001: Parallel Tx Message Processing. This RFC discusses how we can extend the tx parallelization framework to wasm contract execution.
+
+## Background
+
+As discussed in the previous RFC-001, we can prallelize transactions by identifying resource dependencies with variable granularity and building a dependency DAG that allows us to safely process independent transactions concurrently while also maintaining proper ordering for dependent transactions in a block. Wasm contracts are a bit more difficult because they are inherently more opaque in functionality compared to predefined message handlers for chain level messages. As a result, it is more difficult to identify the dependencies of Wasm contracts and parallelize them in the same way as sdk messages.
+
+## Discussion
+
+### Parallelization Approach
+One approach we can take is a gradual parallelization of a wasm contract. When the wasm contract is first uplaoded, we can have it behave as if it uses all resources, essentially making its execution sequential. After the contract has been uploaded (or even as part of upload proposal), we can then allow for a dependency mapping for the contract parallelization which would map between the different execute and query messages for the contract to the resource dependencies that the contract would require during the execution of that execute / query. These dependency mappings can be associated with a contract code ID so that the parallelization can be consistently applied to multiple instances of that contract.
```

Comment:
It might be tricky for contract developers to get this mapping uploaded right. As an intermediate step, I wonder if it's possible to simplify this for them by just asking 1. resources the execute handler itself will query from, 2. resources the execute handler will send messages to update. Then we can construct for them a rough but possibly good enough dependency mapping:
1a. Read from ANY query resource
1b. Write to `wasm/{contract address}`
2. Write to ANY message resource
Note that this might get recursive since a contract can send message to another contract, which is okay as long as there is no cycle (to think about it I'm curious if wasm itself has any guard against it, since even without our parallelization effort, cycles in contract calling can still be a problem as in infinite loop)

Reply:
For contracts with `reply` though I can't think of a good way at the moment. I don't think too many contracts make use of `reply` anyway so maybe we can just make all execute with reply sequential

Reply:
The thing about the proposed approach above is that most contracts will interact with the chain modules outside of the contract itself, with things such as bank transfers, which means that in that case the contract would take control of READ ANY and WRITE ANY, so it would result in basically all contract becoming blocking / sequential inherently. Haven't thought too much about reply, but my thought was that the resource wouldn't be released until AFTER the TX is done executing, which should include the reply processing.

Reply:
It just needs to block on the resource list of the message type it could possibly send, which should be a smaller set than ANY i think

Reply:
yeah thats true, and thats where the accuracy of the mapped dependencies is important, I'd imagine we could work with contract developers to identify the best parallelization dependencies for their contract if theyre having trouble identifying the best approach